### PR TITLE
filterx: use a single compound expr to represent list of expressions

### PR DESCRIPTION
**Pull request by bazsi**
This merges various forms of expression lists into a single FilterXExpr instance:
* true and false branches of conditions
* top-level expression list of filterx blocks
* expression lists generated by our grammar (e.g. generator expressions)

It also implements expression tracing at all of these levels, so instead of only tracing top level statements, it also steps into the true/false branches of if statements.

It also fixes a few other issues:
* the requirement of having to add a "true" as the last statement in ifs to avoid dropping the entire message.
*

### CODE REVIEW

File: lib/filterx/filterx-grammar.ym [hunk 1]
```diff
@@ -473,36 +473,49 @@ conditional
...
 	  }
 	| if KW_ELIF '(' expr ')' codeblock
 	  {
-	    $$ = filterx_conditional_add_false_branch((FilterXConditional*)$1, (FilterXConditional*)filterx_conditional_new_conditional_codeblock($4, $6));
+	    /* create new conditional */
+	    FilterXExpr *elif_expr = filterx_conditional_new($4);
+            filterx_conditional_set_true_branch(elif_expr, $6);
+
+	    /* link it into the else branch of the last if */
+	    filterx_conditional_set_false_branch($1, elif_expr);
```
Comment (alltilla) [inline lib/filterx/filterx-grammar.ym:489]: We must link to the tailing if.

```diff
-	     filterx_conditional_set_false_branch($1, elif_expr);
+            FilterXExpr *tailing_if = filterx_conditional_find_tail($1);
+	     filterx_conditional_set_false_branch(tailing_if, elif_expr);
```

Also an E2E test for multiple `elif` branches would be nice.

File: lib/filterx/expr-condition.c [hunk 2]
```diff
@@ -55,6 +56,26 @@ _eval(FilterXExpr *s)
...
+      GString *buf = scratch_buffers_alloc_and_mark(&mark);
+
+      if (!filterx_object_repr(condition_value, buf))
+        {
+          LogMessageValueType t;
+          if (!filterx_object_marshal(condition_value, buf, &t))
+            g_assert_not_reached();
+        }
+
+      msg_trace("FILTERX  COND",
```
Comment (alltilla) [inline lib/filterx/expr-condition.c:71]: [optional]

I would suggest to code the result to the trace log's message for two reasons:
1. It makes it easier to find a truthy condition.
2. It becomes the same length as and `ESTEP`, `ERROR`.

```diff
-      msg_trace("FILTERX  COND",
+      msg_trace(filterx_object_truthy(condition_value) ? "FILTERX CONDT" : "FILTERX CONDF",
```

**Review by bshifter — COMMENTED** [review]
lgtm

File: lib/filterx/tests/test_expr_compound.c [hunk 3]
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2022 Balazs Scheidler <balazs.scheidler@axoflow.com>
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/filterx/tests/test_expr_compound.c:2]: fixed.

File: lib/filterx/filterx-expr.h [hunk 4]
```diff
@@ -157,6 +157,4 @@ typedef struct _FilterXBinaryOp
 void filterx_binary_op_free_method(FilterXExpr *s);
 void filterx_binary_op_init_instance(FilterXBinaryOp *self, FilterXExpr *lhs, FilterXExpr *rhs);
 
-gboolean filterx_expr_list_eval(GList *expressions, FilterXObject **result);
-
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/filterx/filterx-expr.h:160]: fixed.

File: lib/filterx/expr-compound.c [hunk 5]
```diff
@@ -39,16 +39,44 @@ typedef struct _FilterXCompoundExpr
...
-  if (!(*result))
+  if (!res)
     return FALSE;
 
-  if (!expr->ignore_falsy_result && filterx_object_falsy(*result))
-    return FALSE;
-  return TRUE;
+  gboolean success = expr->ignore_falsy_result || filterx_object_truthy(res);
+
+  if ((!success || trace_flag) && !expr->suppress_from_trace)
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/filterx/expr-compound.c:50]: fixed.

File: lib/filterx/expr-condition.c (hunk 2 above)
Comment (bazsi) [reply lib/filterx/expr-condition.c:71]: good idea, fixed.

File: lib/filterx/filterx-eval.c [hunk 6]
```diff
@@ -188,66 +188,22 @@ filterx_eval_store_weak_ref(FilterXObject *object)
...
-                  evt_tag_str("type", res->type->name));
-    }
-
-  filterx_object_unref(res);
-  return success;
-}
-
 gboolean
-filterx_eval_exec_statements(FilterXEvalContext *context, GList *statements, LogMessage *msg)
+filterx_eval_expr(FilterXEvalContext *context, FilterXExpr *expr, LogMessage *msg)
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/filterx/filterx-eval.c]: I did rename filterx_eval_expr() to filterx_eval_exec() as the last commit (just pushed).

With regards to your 2nd point, exec will happily take any expr and we agreed that we should retain the original prototype.

File: lib/filterx/filterx-eval.c [hunk 7]
```diff
@@ -188,66 +188,22 @@ filterx_eval_store_weak_ref(FilterXObject *object)
...
     {
-      FilterXExpr *expr = l->data;
-      if (!_evaluate_statement(expr))
-        {
-          goto fail;
-        }
+      filterx_eval_clear_errors();
+      goto fail;
     }
+  success = filterx_object_truthy(res);
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/filterx/filterx-eval.c:207]: As discussed in the previous note, we can take something that is not FilterXCompoundExpr so the truthy check is needed.

File: lib/filterx/filterx-grammar.ym (hunk 1 above)
Comment (bazsi) [reply lib/filterx/filterx-grammar.ym]: Added a light test and also fixed the bug in the last iteration.

**Comment by bazsi** [comment]
I've addressed all the review notes.

**Review by alltilla — APPROVED** [review]

**Pull request merged**